scalar_udf.go

```diff
@@ -0,0 +1,152 @@
+package duckdb
+
+// Related issues: https://golang.org/issue/19835, https://golang.org/issue/19837.
+
+/*
+#include <stdlib.h>
+#include <duckdb.h>
+
+void scalar_udf_callback(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+void scalar_udf_delete_callback(void *);
+
+typedef void (*scalar_udf_callback_t)(duckdb_function_info, duckdb_data_chunk, duckdb_vector);
+*/
+import "C"
+
+import (
+	"database/sql"
+	"database/sql/driver"
+	"runtime/cgo"
+	"strings"
+	"unsafe"
+)
+
+/*
+Interface notes.
+- Currently, you must cast your driver.Value's to int32. Casting to just int is not implemented.
+*/
+
+type ScalarFunctionConfig struct {
+	InputTypes []string
+	ResultType string
+}
+
+type ScalarFunction interface {
+	Config() ScalarFunctionConfig
+	ExecuteRow(args []driver.Value) (any, error)
+	SetError(err error)
+}
+
+//export scalar_udf_callback
+func scalar_udf_callback(info C.duckdb_function_info, input C.duckdb_data_chunk, output C.duckdb_vector) {
+	// info is a void* pointer to our ScalarFunction.
+	h := *(*cgo.Handle)(unsafe.Pointer(info))
+	scalarFunction := h.Value().(ScalarFunction)
```

Comment:
This is invalid use of `cgo.Handle`, and it bit me hard last week. `cgo.Handle` is already a pointer, no need to wrap another pointer around it. In fact you can sometimes get cgo errors when using it with a pointer like this, which really confused me at first.
You probably copied this from me so just letting you know before you get any issues with this :)

Reply:
I copied it from their (Go) documentation example, but I am confused, so I am no saying it's not wrong, haha. I figured the second example applies, as `info` is a `void *` pointer?
```C
//! Additional function info. When setting this info, it is necessary to pass a destroy-callback function.
typedef struct _duckdb_function_info {
	void *__val;
} * duckdb_function_info;
```

```Go
// Handle provides a way to pass values that contain Go pointers
// (pointers to memory allocated by Go) between Go and C without
// breaking the cgo pointer passing rules. A Handle is an integer
// value that can represent any Go value. A Handle can be passed
// through C and back to Go, and Go code can use the Handle to
// retrieve the original Go value.
//
// The underlying type of Handle is guaranteed to fit in an integer type
// that is large enough to hold the bit pattern of any pointer. The zero
// value of a Handle is not valid, and thus is safe to use as a sentinel
// in C APIs.
//
// For instance, on the Go side:
//
//	package main
//
//	/*
//	#include <stdint.h> // for uintptr_t
//
//	extern void MyGoPrint(uintptr_t handle);
//	void myprint(uintptr_t handle);
//	*/
//	import "C"
//	import "runtime/cgo"
//
//	//export MyGoPrint
//	func MyGoPrint(handle C.uintptr_t) {
//		h := cgo.Handle(handle)
//		val := h.Value().(string)
//		println(val)
//		h.Delete()
//	}
//
//	func main() {
//		val := "hello Go"
//		C.myprint(C.uintptr_t(cgo.NewHandle(val)))
//		// Output: hello Go
//	}
//
// and on the C side:
//
//	#include <stdint.h> // for uintptr_t
//
//	// A Go function
//	extern void MyGoPrint(uintptr_t handle);
//
//	// A C function
//	void myprint(uintptr_t handle) {
//	    MyGoPrint(handle);
//	}
//
// Some C functions accept a void* argument that points to an arbitrary
// data value supplied by the caller. It is not safe to coerce a [cgo.Handle]
// (an integer) to a Go [unsafe.Pointer], but instead we can pass the address
// of the cgo.Handle to the void* parameter, as in this variant of the
// previous example:
//
//	package main
//
//	/*
//	extern void MyGoPrint(void *context);
//	static inline void myprint(void *context) {
//	    MyGoPrint(context);
//	}
//	*/
//	import "C"
//	import (
//		"runtime/cgo"
//		"unsafe"
//	)
//
//	//export MyGoPrint
//	func MyGoPrint(context unsafe.Pointer) {
//		h := *(*cgo.Handle)(context)
//		val := h.Value().(string)
//		println(val)
//		h.Delete()
//	}
//
//	func main() {
//		val := "hello Go"
//		h := cgo.NewHandle(val)
//		C.myprint(unsafe.Pointer(&h))
//		// Output: hello Go
//	}
type Handle uintptr
```

Reply:
huh, that is a good point. I will revert this in my PR and try again but I got some errors. Maybe I should open a bug report for go? I definitely got `...invalid handle...` errors when originally, but I will see if I can find out whats wrong.

Reply:
See https://github.com/golang/go/issues/68044 for what was happening. We should pin(and unpin) the handle before handing it off and after deleting it

Reply:
Thanks a lot for diving into this and finding the proper approach!